### DISCUSSION TITLE
Know more about device SDK

### DISCUSSION THREAD

**Discussion by rsl-vitragshah** [aws/aws-iot-device-sdk-js-v2#441] [Q&A] [answered]
I am currently confused on which SDK to use for below purpose:

1. To setup an IOT device on my MacOS which will interact with AWS IOT
2. To setup an node backend project which will interact with AWS IOT

I tried to find documentations but there is not clear differentiation between above scenarios.

**Comment by jmklix** [comment]
> 1. To setup an IOT device on my MacOS which will interact with AWS IOT

For this you will need to use the main [aws-sdk-js-v3](https://github.com/aws/aws-sdk-js-v3) . Check out the API docs [here](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/iot/) to see how to setup IoT devices on AWS. I would also recommend learning how to manually set up devices in the [AWS console](https://docs.aws.amazon.com/iot/latest/developerguide/iot-gs.html) first if you don't already know how

> 2. To setup an node backend project which will interact with AWS IOT

Use this sdk to run in a node backend project to interact with AWS IoT. Check out our [samples](https://github.com/aws/aws-iot-device-sdk-js-v2/blob/main/samples/node/pub_sub/README.md) to see how to use this sdk. 

Please let me know if you have any further questions or need more clarification for how to use these sdks.

(edit: You can also use [aws-sdk-js](https://github.com/aws/aws-sdk-js) for both but this is being moved to maintenance mode soon)

**Comment by rsl-vitragshah** [comment]
There are a lot SDKs with similar names. Can you help clarify usage for each of them and how they plan role w.r.t AWS IOT core both from IOT device and users apps (web, mobile).

https://github.com/aws/aws-sdk-js-v3
https://www.npmjs.com/package/aws-iot-device-sdk-v2
https://www.npmjs.com/package/aws-sdk

**Reply by rsl-vitragshah** [reply]
In my node project I used: https://www.npmjs.com/package/aws-sdk to connect with AWS IOT core. `new AWS.IotData`

To setup IOT hardware device I downloaded connect-device-kit for Python which is seems to be using "aws-iot-device-sdk-python-v2". This is opposite to details provided by you for #2 answer

**Reply by jmklix** [reply]
[aws-sdk-js-v3](https://github.com/aws/aws-sdk-js-v3):
  - This is the newest version of the js sdk that is used for interacting with aws services
  - This is modularized (i.e. @aws-sdk/client-iot)
  - This sdk does not have MQTT operations build into it
  - Use [client-iot](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/client/iot/) for setup/edit of IoT devices
  
[aws-sdk](https://github.com/aws/aws-sdk-js) (not recommended to use this)
  - This is the v2 version of the js sdk and will be going into maintenance mode soon. Because of that I would not recommend using this.
  - This sdk allows both setup and usage of IoT devices, but it is not optimized for IoT devices
  - Use [Iot](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/Iot.html) for setup/edit of IoT devices
  - Use [IotData](https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/IotData.html) for MQTT calls 
  
[aws-iot-device-sdk-js-v2](https://github.com/aws/aws-iot-device-sdk-js-v2)
  - This is the v2 version of the IoT sdk and is completely separate from the above js sdk's
  - This was specifically written for IoT devices and is optimized for MQTT operations.
  - Use this sdk to communication with IoT core (check out the [pubsub](https://github.com/aws/aws-iot-device-sdk-js-v2/blob/main/samples/node/pub_sub/README.md) for basic usage)
  - In reference to your above comment, this sdk is not actually setting up the IoT device. Rather going through the steps on the AWS console to download the connection-device-kit is setting up the IoT device for you.
  
Please let me know if you would like me to clarify any more details about these sdk's

**Reply by rsl-vitragshah** [reply]
Thanks for detailed reply. 
To summarize, both [aws-sdk-js-v3](https://github.com/aws/aws-sdk-js-v3) and [aws-iot-device-sdk-v2](https://github.com/aws/aws-iot-device-sdk-js-v2) will be used in backend project to perform different operations related to IOT Core.

One more question very specific to IOT hardware device: Consider I am programming an IOT hardware device which will interact with AWS IOT (pub,sub through MQTT). For this case which SDK will be used?
Based on what I understood till now - same SDKs will be used for both hardware device and backend project.

**Reply by rsl-vitragshah** [reply]
Also it would be really helpful if you can share samples links which use latest expected AWS IOT SDKs for nodejs, web app and mobile apps.

**Reply by jmklix** [reply, marked as answer]
Q: Consider I am programming an IOT hardware device which will interact with AWS IOT (pub,sub through MQTT). For this case which SDK will be used?

A: For this you should use the aws-iot-device-sdk-js-v2 sdk. (You can also choose to use any of the other device sdk's. We have them for the following languages: [Java](https://github.com/aws/aws-iot-device-sdk-java-v2), [Python](https://github.com/aws/aws-iot-device-sdk-python-v2), and [C++](https://github.com/aws/aws-iot-device-sdk-cpp-v2))

You can find all of the latest IoT samples [here](https://github.com/aws/aws-iot-device-sdk-js-v2/tree/main/samples). The samples are divided into Node and Browser folders for nodejs and web apps respectively.
